lua/barbar/layout.lua

```diff
@@ -20,130 +20,159 @@ local state = require'barbar.state'
 --- The number of sides of each buffer in the tabline.
 local SIDES_OF_BUFFER = 2
 
+--- The length of a slash (`#'/'`)
+local SLASH_LEN = #'/'
+
+--- The length of one space (`#' '`)
+local SPACE_LEN = #' '
+
 --- @class barbar.layout.data
---- @field actual_width integer
---- @field available_width integer
---- @field base_widths integer[]
---- @field buffers_width integer
---- @field padding_width integer
---- @field tabpages_width integer
---- @field used_width integer
+--- @field actual_width integer the sum of the `base_widths` plus the `padding_width` allocated to each buffer
+--- @field base_widths integer[] the minimum amount of space taken up by each buffer
+--- @field buffers_width integer the amount of space available to be taken up by buffers
+--- @field padding_width integer the amount of padding used on each side of each buffer
+--- @field scroll_max integer the maximum position which can be scrolled to
+--- @field tabpages_width integer the amount of space taken up by the tabpage indicator
 
 --- @class barbar.Layout
 --- @field buffers integer[] different from `state.buffers` in that the `hide` option is respected. Only updated when calling `calculate_buffers_width`.
 local Layout = {buffers = {}}
 
---- The number of characters needed to represent the tabpages.
---- @return integer width
-function Layout.calculate_tabpages_width()
-  local current = tabpagenr()
-  local total   = tabpagenr('$')
-  if not config.options.tabpages or total == 1 then
-    return 0
-  end
-  return 1 + tostring(current):len() + 1 + tostring(total):len() + 1
+--- Calculate the current layout of the bufferline.
+--- @return barbar.layout.data
+function Layout.calculate()
```

Comment:
Before you do more work here, wdyt if we store the pinned buffers in a separate `state.buffers_pinned` list? I think it could simplify a bit our state management and our layout logic.

OTOH, I'm not sure if that would make pinning animations harder to implement.

I'd be happy to hear your thoughts on this.

Reply:
I did try to implement it that way, but it ended up complicating things:

* Buffer navigation: we iterate over `state.buffers` for choosing the next/previous buffer. We would have to check two lists now.
* Increased time-complexity for `state.is_pinned`: right now it is (presumably) `O(log n)` since it resolves to `state.data_by_bufnr[bufnr].pinned`. To re-implement that functionality we'd need to either:
    1. `utils.index_of(state.pinned_buffers, bufnr)` which is `O(n)` worst case
    2. Keep `pinned` in `data_by_bufnr`, but now we have to keep those items in sync.

In all, anywhere `state.buffers` or `Layout.buffers` is referenced, we'd have to update to check two lists (plus bounding for when going backwards past the beginning, or forwards past the end). 

It would let us remove `sort_pins_left`, but we might be able to do that now anyway since pinned buffers are rendered statically (I'll double check)

Reply:
Your points make total sense, let's keep just `state.buffers` then.

But maybe during the layout/render phase we should split them in 2 lists? Cuz I'm thinking that we're basically going to have 2 sections, a static one for pinned buffers and a scrollable one for unpinned buffers. Let me know what you think. My main concern is that we have readable & understandeable code, tbh sometimes I get lost in our layout code ^^

Reply:
I do that in `generate_tabline` on (line 533 is where the var is defined). Pinned items are appended to the pinned item list, others go in the general list (which makes doing converting pinned buffers statically much easier).

In layout I keep `base_widths` together in one list for mostly the same reasons I listed earlier (though I did try separating the widths out). However I separately calculate the `pinned_width_sum` and `pinned_count` as they are necessary values for determining the size of the pinned buffer section